runtime/compiler/control/JitDump.cpp

```diff
@@ -428,35 +429,38 @@ runJitdump(char *label, J9RASdumpContext *context, J9RASdumpAgent *agent)
-         const char *name;
-         void *value;
-         U_32 infoType = j9sig_info(crashedThread->gpInfo, J9PORT_SIG_CONTROL, J9PORT_SIG_CONTROL_PC, &name, &value);
-
-         if (J9PORT_SIG_VALUE_ADDRESS == infoType)
+         if (NULL != crashedThread->gpInfo)
             {
-            J9JITExceptionTable *metadata = jitConfig->jitGetExceptionTableFromPC(crashedThread, *reinterpret_cast<UDATA*>(value));
-            if (NULL != metadata)
+            // The stack walker may not be able to walk the stack if the crash did not happen on a transition frame. In
+            // such cases the stack walker will resume walking the stack on the last known valid point, which will be a
+            // transition frame further up the stack (ex. INT -> JIT transition frame). This will result in the stack
+            // walker potentially skipping some JIT methods on the backtrace. This is not desirable. Chances are high
+            // that the crash happen because of a miscompilation in the first JIT compiled method on the stack, so it
+            // is imperative that we recompile the method we actually crashed in.
+            const char *name;
+            void *value;
+            U_32 infoType = j9sig_info(crashedThread->gpInfo, J9PORT_SIG_CONTROL, J9PORT_SIG_CONTROL_PC, &name, &value);
+
+            if (J9PORT_SIG_VALUE_ADDRESS == infoType)
                {
-               auto *bodyInfo = reinterpret_cast<TR_PersistentJittedBodyInfo *>(metadata->bodyInfo);
-               if (NULL != bodyInfo)
+               J9JITExceptionTable *metadata = jitConfig->jitGetExceptionTableFromPC(crashedThread, *reinterpret_cast<UDATA*>(value));
+               if (NULL != metadata)
                   {
-                  jitDumpRecompileWithTracing(
-                     crashedThread,
-                     metadata->ramMethod,
-                     compInfo,
-                     bodyInfo->getHotness(),
-                     bodyInfo->getIsProfilingBody(),
-                     NULL,
-                     bodyInfo->getIsAotedBody(),
-                     bodyInfo->getStartPCAfterPreviousCompile(),
-                     jitdumpFile
-                  );
+                  auto *bodyInfo = reinterpret_cast<TR_PersistentJittedBodyInfo *>(metadata->bodyInfo);
+                  if (NULL != bodyInfo)
+                     {
+                     jitDumpRecompileWithTracing(
+                        crashedThread,
+                        metadata->ramMethod,
+                        compInfo,
+                        bodyInfo->getHotness(),
+                        bodyInfo->getIsProfilingBody(),
+                        NULL,
+                        bodyInfo->getIsAotedBody(),
+                        bodyInfo->getStartPCAfterPreviousCompile(),
+                        jitdumpFile
+                     );
+                     }
                   }
                }
             }
```

Comment:
What is the purpose of the code that follows below? We end up calling `crashedThread->javaVM->walkStackFrames(crashedThread, &walkState);` but to what end?

Reply:
It will walk the stack calling the callback function we registered (`jitDumpStackFrameIterator`). This function will then recompile all JIT methods encountered on the stack. This can be useful if the problem originates a few frames above the crashed method, and we are able to walk the entire stack frame.

Typical example of this is we call a VM helper from inside JITted code and we crash somewhere in the VM (ex. bad argument passed). We should then be able to walk the entire stack and recompile all JIT methods.